### ISSUE TITLE
Add the ability to drag custom price line and change its price

### ISSUE THREAD

**Issue by randalhsu** [tradingview/lightweight-charts#1086]
**Is your feature request related to a problem? Please describe.**

No.

**Describe the solution you'd like**

Able to drag custom price line.
[Demo](https://jsfiddle.net/randalhsu/5un8r93g/) (The orders can be dragged in the price scale.)

I already have a [patch](https://github.com/randalhsu/ChartExerciser/blob/main/lightweight-charts-patch/0002-Draggable-custom-price-line.patch) implementing this.
Might take some efforts to rebase and write test case etc. if this feature request is approved.

**Additional context**

Proposal:
1. Add boolean parameter `draggable` to [createPriceLine()](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/ISeriesApi#createpriceline).
2. Able to get "the price before dragging". I think it should be stored as a string to preserve precisions? User can parse it later if needed.
3. Change cursor when hovering over a draggable price line or dragging.

**Comment by julio899** [comment]
That's amazing how I can help?

**Comment by shaunbeh** [comment]
@randalhsu hi and tnx for ur work
I really need this feature and was hoping to have the library code till this gets approved,
u have provided the production ready file in ur demo and patch files and can't run npm run build because there is no subscribeCustomPriceLineDragged function in the main library, what should I do to have this in a react project?

tnx and sorry for the stupid question I have asked, Im new to this library tweeking library files

**Comment by randalhsu** [comment]
@shaunbeh 
Manually build `lightweight-charts` with the patch:
```
git clone https://github.com/tradingview/lightweight-charts.git
cd lightweight-charts
git checkout -b v3.8.0 tags/v3.8.0
wget https://github.com/randalhsu/OPAL/raw/main/lightweight-charts-patch/0002-Draggable-custom-price-line.patch
git am 0002-Draggable-custom-price-line.patch
npm install --legacy-peer-deps
npm run build:prod
```
You need to use this file: `dist/lightweight-charts.standalone.production.js`

**Comment by mr-smit** [comment]
Thanks for your work, was thinking to implement this, but you already did it. thanks.

**Comment by julio899** [comment]
> @shaunbeh Manually build `lightweight-charts` with the patch:
> 
> ```
> git clone https://github.com/tradingview/lightweight-charts.git
> cd lightweight-charts
> git checkout -b v3.8.0 tags/v3.8.0
> wget https://github.com/randalhsu/OPAL/raw/main/lightweight-charts-patch/0002-Draggable-custom-price-line.patch
> git am 0002-Draggable-custom-price-line.patch
> npm install --legacy-peer-deps
> npm run build:prod
> ```
> 
> You need to use this file: `dist/lightweight-charts.standalone.production.js`

Thanks so much I appreciate a lot.

**Comment by felipecsl** [comment]
Is there an existing pull request with this change? I couldn't find any. If not, I'd be happy to put one together since I need this change as well

**Comment by felipecsl** [comment]
Code is here https://github.com/felipecsl/lightweight-charts/compare/master...felipecsl:lightweight-charts:feat--drag-price-line
Keep in mind that it also includes changes from #1214 which are also present in my fork of the library.
I pushed it to npmjs.org [here](https://www.npmjs.com/package/@felipecsl/lightweight-charts)
Once #1214 is merged I can send a separate pull request for this change

**Comment by randalhsu** [comment]
@felipecsl There's no pull request yet. Thanks for helping, hope it gets merged soon.

**Comment by jussirantala** [comment]
> Code is here [felipecsl/lightweight-charts@master...felipecsl:lightweight-charts:feat--drag-price-line](https://github.com/felipecsl/lightweight-charts/compare/master...felipecsl:lightweight-charts:feat--drag-price-line) Keep in mind that it also includes changes from #1214 which are also present in my fork of the library. I pushed it to npmjs.org [here](https://www.npmjs.com/package/@felipecsl/lightweight-charts) Once #1214 is merged I can send a separate pull request for this change

#1214 is now merged :)

**Comment by kevinkga** [comment]
> Code is here [felipecsl/lightweight-charts@master...felipecsl:lightweight-charts:feat--drag-price-line](https://github.com/felipecsl/lightweight-charts/compare/master...felipecsl:lightweight-charts:feat--drag-price-line) Keep in mind that it also includes changes from #1214 which are also present in my fork of the library. I pushed it to npmjs.org [here](https://www.npmjs.com/package/@felipecsl/lightweight-charts) Once #1214 is merged I can send a separate pull request for this change

I just tried the latest version (4.0.7) of your fork but chart.subscribeCustomPriceLineDragged seems to be missing. Am I looking at the right version?

**Comment by felipecsl** [comment]
@kevinkga yeah that's not the right version. You'll have to grab an [older version](https://www.npmjs.com/package/@felipecsl/lightweight-charts?activeTab=versions) to pick up that change, but those will be a bit outdated. I can probably sync it with the latest and publish a new version with it

**Comment by jussirantala** [comment]
They will never add this feature. They want to sell the advanced library which has trading panel and costs 14k/year.

**Comment by darwin403** [comment]
@SlicedSilver can you please consider this in one of the future releases. This would be great for order placement mechanisms.

**Comment by brlocky** [comment]
@felipecsl 

thank you for this amazing feature.


I confirm that version 4.0.4 is working fine. 4.0.5 and 4.0.6 are not
  `  "@felipecsl/lightweight-charts": "4.0.4",`

**Comment by SlicedSilver** [comment]
Thank you for your feature request for Lightweight Charts. In order to keep the library lightweight and maintain a narrow focus, we strive to include only strictly required features in the core library. As a result, we have decided that this requested feature should be implemented as a plugin instead, and will be closing this issue as a result.

We recently introduced the ability to write plugins for Lightweight Charts, allowing users like you to extend the functionality of the library without adding it to the core. We highly recommend implementing your feature request as a plugin to achieve the desired functionality while keeping the core bundle size smaller for all users.

To get started with plugins, please refer to our [Plugins Documentation](https://tradingview.github.io/lightweight-charts/docs/plugins/intro) for a better understanding of what is possible and how plugins work. You can also explore our collection of [plugin examples](https://github.com/tradingview/lightweight-charts/tree/master/plugin-examples) (with a [preview hosted here](https://tradingview.github.io/lightweight-charts/plugin-examples/)) for inspiration and guidance on implementing specific functionality.

For your particular feature request, we suggest taking a look at the following examples (see above links):
- User Defined Price Lines
- User Price Alerts

To help you get started quickly, we have created an NPM package called [create-lwc-plugin](https://www.npmjs.com/package/create-lwc-plugin), which sets up a plugin project for you. This way, you can hit the ground running with your plugin development.

If you do create a plugin based on your feature request, we would appreciate it if you could publish it on GitHub or NPM and let us know. This way, we can share it with the rest of the community. It would be helpful if you could prefix your NPM package name with `lwc-plugin-` for easier discoverability by others.

Thank you for your understanding and support. We believe implementing your feature request as a plugin will greatly benefit both you and the Lightweight Charts community.

**Comment by thichmautim** [comment]
> @shaunbeh Manually build `lightweight-charts` with the patch:
> 
> ```
> git clone https://github.com/tradingview/lightweight-charts.git
> cd lightweight-charts
> git checkout -b v3.8.0 tags/v3.8.0
> wget https://github.com/randalhsu/OPAL/raw/main/lightweight-charts-patch/0002-Draggable-custom-price-line.patch
> git am 0002-Draggable-custom-price-line.patch
> npm install --legacy-peer-deps
> npm run build:prod
> ```
> 
> You need to use this file: `dist/lightweight-charts.standalone.production.js`

Could you please upload "dist/lightweight-charts.standalone.production.js". I was built but get some errors.

**Comment by thichmautim** [comment]
> @felipecsl
> 
> thank you for this amazing feature.
> 
> I confirm that version 4.0.4 is working fine. 4.0.5 and 4.0.6 are not ` "@felipecsl/lightweight-charts": "4.0.4",`

Could you please upload "dist/lightweight-charts.standalone.production.js". I was built but get some errors.

**Comment by levino** [comment]
Was the plugin ever written?

**Comment by mr-smit** [comment]
> levino

https://github.com/tradingview/lightweight-charts/tree/master/plugin-examples/src/plugins

**Comment by levino** [comment]
I meant a plugin to drag a custom price line and change its price. I do not understand which plugin in the list would provide the functionality.

**Comment by alexxvedo** [comment]
I was looking for the plugin too but couldn't find it so I created a little demo of how it could be implemented using only eventListeners.

Demo: [codi.link Demo Drag PriceLine](https://codi.link/PCFET0NUWVBFIGh0bWw+CjxodG1sIGxhbmc9ImVuIj4KPGhlYWQ+CiAgPG1ldGEgY2hhcnNldD0iVVRGLTgiPgogIDxtZXRhIG5hbWU9InZpZXdwb3J0IiBjb250ZW50PSJ3aWR0aD1kZXZpY2Utd2lkdGgsIGluaXRpYWwtc2NhbGU9MS4wIj4KICA8dGl0bGU+TGlnaHR3ZWlnaHQgQ2hhcnRzIERlbW88L3RpdGxlPgogIDxzdHlsZT4KICAgICNjaGFydCB7CiAgICAgIHdpZHRoOiAxMDAlOwogICAgICBoZWlnaHQ6IDUwMHB4OwogICAgfQogIDwvc3R5bGU+CjwvaGVhZD4KPGJvZHk+CiAgPGRpdiBpZD0iY2hhcnQiPjwvZGl2PgogIDxzY3JpcHQgc3JjPSJodHRwczovL3VucGtnLmNvbS9saWdodHdlaWdodC1jaGFydHMvZGlzdC9saWdodHdlaWdodC1jaGFydHMuc3RhbmRhbG9uZS5wcm9kdWN0aW9uLmpzIj48L3NjcmlwdD4KICA8c2NyaXB0IHNyYz0ic2NyaXB0LmpzIj48L3NjcmlwdD4KPC9ib2R5Pgo8L2h0bWw+Cg==%7C%7CIGNvbnN0IGNoYXJ0ID0gTGlnaHR3ZWlnaHRDaGFydHMuY3JlYXRlQ2hhcnQoZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoJ2NoYXJ0JyksIHsKICAgICAgd2lkdGg6IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdjaGFydCcpLmNsaWVudFdpZHRoLAogICAgICBoZWlnaHQ6IDUwMCwKICAgICAgbGF5b3V0OiB7CiAgICAgICAgYmFja2dyb3VuZENvbG9yOiAnI2ZmZmZmZicsCiAgICAgICAgdGV4dENvbG9yOiAnIzAwMCcsCiAgICAgIH0sCiAgICAgIGdyaWQ6IHsKICAgICAgICB2ZXJ0TGluZXM6IHsKICAgICAgICAgIGNvbG9yOiAnI2UxZTFlMScsCiAgICAgICAgfSwKICAgICAgICBob3J6TGluZXM6IHsKICAgICAgICAgIGNvbG9yOiAnI2UxZTFlMScsCiAgICAgICAgfSwKICAgICAgfSwKICAgICAgcHJpY2VTY2FsZTogewogICAgICAgIGJvcmRlckNvbG9yOiAnI2NjYycsCiAgICAgIH0sCiAgICAgIHRpbWVTY2FsZTogewogICAgICAgIGJvcmRlckNvbG9yOiAnI2NjYycsCiAgICAgIH0sCiAgICB9KTsKCiAgICBjb25zdCBjYW5kbGVzdGlja1NlcmllcyA9IGNoYXJ0LmFkZENhbmRsZXN0aWNrU2VyaWVzKHsKICAgICAgdXBDb2xvcjogJyM0Y2FmNTAnLAogICAgICBkb3duQ29sb3I6ICcjZjQ0MzM2JywKICAgICAgYm9yZGVyVmlzaWJsZTogZmFsc2UsCiAgICAgIHdpY2tVcENvbG9yOiAnIzRjYWY1MCcsCiAgICAgIHdpY2tEb3duQ29sb3I6ICcjZjQ0MzM2JywKICAgIH0pOwoKICAgIGNhbmRsZXN0aWNrU2VyaWVzLnNldERhdGEoWwogICAgICB7IHRpbWU6ICcyMDIyLTAxLTAxJywgb3BlbjogMTAwLCBoaWdoOiAxMTAsIGxvdzogOTAsIGNsb3NlOiAxMDUgfSwKICAgICAgeyB0aW1lOiAnMjAyMi0wMS0wMicsIG9wZW46IDEwNSwgaGlnaDogMTE1LCBsb3c6IDk1LCBjbG9zZTogMTEwIH0sCiAgICAgIHsgdGltZTogJzIwMjItMDEtMDMnLCBvcGVuOiAxMTAsIGhpZ2g6IDEyMCwgbG93OiAxMDAsIGNsb3NlOiAxMTUgfSwKICAgIF0pOwoKICAgIGNvbnN0IHByaWNlTGluZXMgPSBbCiAgICAgIGNhbmRsZXN0aWNrU2VyaWVzLmNyZWF0ZVByaWNlTGluZSh7CiAgICAgICAgcHJpY2U6IDExMCwKICAgICAgICBjb2xvcjogJ3JlZCcsCiAgICAgICAgbGluZVdpZHRoOiAyLAogICAgICAgIGxpbmVTdHlsZTogTGlnaHR3ZWlnaHRDaGFydHMuTGluZVN0eWxlLlNvbGlkLAogICAgICAgIGF4aXNMYWJlbFZpc2libGU6IHRydWUsCiAgICAgICAgdGl0bGU6ICdMaW5lIDEnLAogICAgICB9KSwKICAgICAgY2FuZGxlc3RpY2tTZXJpZXMuY3JlYXRlUHJpY2VMaW5lKHsKICAgICAgICBwcmljZTogMTE1LAogICAgICAgIGNvbG9yOiAnYmx1ZScsCiAgICAgICAgbGluZVdpZHRoOiAyLAogICAgICAgIGxpbmVTdHlsZTogTGlnaHR3ZWlnaHRDaGFydHMuTGluZVN0eWxlLlNvbGlkLAogICAgICAgIGF4aXNMYWJlbFZpc2libGU6IHRydWUsCiAgICAgICAgdGl0bGU6ICdMaW5lIDInLAogICAgICB9KSwKICAgIF07CgogICAgbGV0IGRyYWdnaW5nID0gZmFsc2U7CiAgICBsZXQgZHJhZ0luZGV4ID0gbnVsbDsKCiAgICBmdW5jdGlvbiBoYW5kbGVNb3VzZU1vdmUoZXZlbnQpIHsKICAgICAgaWYgKCFkcmFnZ2luZyB8fCBkcmFnSW5kZXggPT09IG51bGwpIHJldHVybjsKCiAgICAgIGNvbnN0IHJlY3QgPSBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCgnY2hhcnQnKS5nZXRCb3VuZGluZ0NsaWVudFJlY3QoKTsKICAgICAgY29uc3QgcHJpY2UgPSBjYW5kbGVzdGlja1Nlcmllcy5jb29yZGluYXRlVG9QcmljZShldmVudC5jbGllbnRZIC0gcmVjdC50b3ApOwoKICAgICAgcHJpY2VMaW5lc1tkcmFnSW5kZXhdLmFwcGx5T3B0aW9ucyh7CiAgICAgICAgcHJpY2U6IHByaWNlLAogICAgICB9KTsKICAgIH0KCiAgICBmdW5jdGlvbiBoYW5kbGVNb3VzZURvd24oZXZlbnQpIHsKICAgICAgY29uc3QgcmVjdCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdjaGFydCcpLmdldEJvdW5kaW5nQ2xpZW50UmVjdCgpOwogICAgICBjb25zdCBwcmljZSA9IGNhbmRsZXN0aWNrU2VyaWVzLmNvb3JkaW5hdGVUb1ByaWNlKGV2ZW50LmNsaWVudFkgLSByZWN0LnRvcCk7CgogICAgICBwcmljZUxpbmVzLmZvckVhY2goKHByaWNlTGluZSwgaW5kZXgpID0+IHsKICAgICAgICAKICAgICAgICBjb25zdCBwcmljZUxpbmVQcmljZSA9IHBhcnNlRmxvYXQocHJpY2VMaW5lLm9wdGlvbnMoKS5wcmljZS50b0ZpeGVkKDUpKTsKICAgICAgICBjb25zdCByb3VuZGVkUHJpY2UgPSBwYXJzZUZsb2F0KHByaWNlLnRvRml4ZWQoNSkpOwoKICAgICAgICBpZiAoTWF0aC5hYnMocm91bmRlZFByaWNlIC0gcHJpY2VMaW5lUHJpY2UpIDwgMC4wMSpwcmljZUxpbmVQcmljZSkgewogICAgICAgICAgZHJhZ2dpbmcgPSB0cnVlOwogICAgICAgICAgZHJhZ0luZGV4ID0gaW5kZXg7CiAgICAgICAgICBjaGFydC5hcHBseU9wdGlvbnMoewogICAgICAgICAgICBoYW5kbGVTY3JvbGw6IGZhbHNlLAogICAgICAgICAgICBoYW5kbGVTY2FsZTogZmFsc2UsCiAgICAgICAgICB9KTsKICAgICAgICB9CiAgICAgIH0pOwogICAgfQoKICAgIGZ1bmN0aW9uIGhhbmRsZU1vdXNlVXAoKSB7CiAgICAgIGlmIChkcmFnSW5kZXggIT09IG51bGwpIHsKICAgICAgICBhbGVydChgUHJpY2UgbGluZSAke2RyYWdJbmRleCsxfSBtb3ZlZCB0bzogYCArIHByaWNlTGluZXNbZHJhZ0luZGV4XS5vcHRpb25zKCkucHJpY2UudG9GaXhlZCg1KSk7CiAgICAgIH0KCiAgICAgIGRyYWdnaW5nID0gZmFsc2U7CiAgICAgIGRyYWdJbmRleCA9IG51bGw7CiAgICAgIGNoYXJ0LmFwcGx5T3B0aW9ucyh7CiAgICAgICAgaGFuZGxlU2Nyb2xsOiB0cnVlLAogICAgICAgIGhhbmRsZVNjYWxlOiB0cnVlLAogICAgICB9KTsKICAgIH0KCiAgICBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCgnY2hhcnQnKS5hZGRFdmVudExpc3RlbmVyKCdtb3VzZW1vdmUnLCBoYW5kbGVNb3VzZU1vdmUpOwogICAgZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoJ2NoYXJ0JykuYWRkRXZlbnRMaXN0ZW5lcignbW91c2Vkb3duJywgaGFuZGxlTW91c2VEb3duKTsKICAgIGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdjaGFydCcpLmFkZEV2ZW50TGlzdGVuZXIoJ21vdXNldXAnLCBoYW5kbGVNb3VzZVVwKTs=)

Hope it helps!
